feat: dollars off discount support in basket/checkout

arslanashraf7 opened the pull request:
#### Pre-Flight checklist

- [ ] Screenshots and design review for any changes that affect layout or styling
  - [ ] Desktop screenshots
  - [ ] Mobile width screenshots
- [ ] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
#2448

#### What's this PR do?
- Adds support for dollars-off coupons on the basket
- Refactors multiple checkout pages to support dollars-off discounts
- Adds dollars-off support for HubSpot
- Refactors and adds tests

#### How should this be manually tested?
This might need to be tested extensively. It might be a good idea to read the conversation on https://github.com/mitodl/mitxpro/issues/2443 before starting the testing.
- Create a couple of coupons 'dollars-off' or 'percent-off; either from Django Admin or through `http://xpro.odl.local:8053/admin/`
- Apply those coupons on an item in the basket (Make sure that all the cases are tested e.g. Global or course-based coupons work exactly as they should for both `dollars-off` and `percent-off`)
- Create dollars-off and percent-off coupons globally and test the case where we pick the best coupon automatically
- Check that the discounts and price displayed to the user is correct
- Disable coupons and check if they become invalid for both `dollars-off` and `percent-off`
- Check the Hubspot that the orders are created successfully there and also they contain the right discount and amounts as mentioned in https://github.com/mitodl/mitxpro/issues/2443
- For Hubspot test keys (You can ask me)
- Also check that the discount or price values never show up to be `negative` since we have dollars-off support now and with dollars, the values can be negative
- No `dollars-off` support for B2b coupons are only percent based and work as expacted and coupons generated through b2b have `discount_type` to be `percent-off`
- No dollars-off support for coupons through sheets so they should work as expacted and have `discount_type` to be `percent-off`.

#### Where should the reviewer start?
Checkout

#### Any background context you want to provide?
- This is part of the epic and for that, the related work is also done in https://github.com/mitodl/mitxpro/pull/2449 and https://github.com/mitodl/mitxpro/pull/2451

annagav commented:
Functionality wise works as expected. The only thing is in Hubspot the orders are getting created successfully, but I can view only the total amount of the order and no discount amount.

arslanashraf7 commented:
> Functionality wise works as expected. The only thing is in Hubspot the orders are getting created successfully, but I can view only the total amount of the order and no discount amount.

Did you try the following steps?
1. Open an order
2. Click `View All properties`
3. And then try checking the values under the section named `Coupon`?

arslanashraf7 commented:
@annagav I save some room for improvement in the code and refactored some bits in https://github.com/mitodl/mitxpro/pull/2455/commits/157224cf75605ba331af28bf85cb1cd534bb0af8. Let me know what you think. 

I also added steps to check the discount amount in HubSpot.

asadali145 commented on ecommerce/serializers.py:
```diff
@@ -258,6 +259,11 @@ def get_amount(self, instance):
+    def get_discount_type(self, instance):
+        """ Get the coupon discount amount """
+        # decimal fields should be represented as strings to prevent floating point parsing problems
+        return str(latest_coupon_version(instance.coupon).payment_version.discount_type)
```
Any specific reason for calling str?

arslanashraf7 replied:
I think we can remove that conversion, I'll change that in the next commit.

arslanashraf7 replied:
Addressed in https://github.com/mitodl/mitxpro/pull/2455/commits/57ef368881d053254364742e56e270d4dad67fc6

asadali145 commented on ecommerce/utils.py:
```diff
@@ -103,3 +103,8 @@ def validate_amount(discount_type, amount):
         return "The amount should be between (0 - 1) when discount type is percent-off."
+
+
+def positive_or_zero(number):
+    """Return 0 if a number is negative otherwise return number"""
+    return 0 if number < 0 else number
```
Can we use max(0, number) instead of creating this function?

arslanashraf7 replied:
Intentionally didn't do that, We are just comparing two values and I believe the max function would be slower than the ternary operator.

asadali145 reviewed (commented):
Code LGTM! A couple of minor suggestions. I will test the changes and post a comment after that.

arslanashraf7 commented on ecommerce/api.py:
```diff
@@ -299,12 +305,32 @@ def get_valid_coupon_versions(
+
+            global_coupon_version_subquery = global_coupon_version_subquery.filter(
+                payment_version__amount=decimal.Decimal(1)
+            )
```
Addressed in https://github.com/mitodl/mitxpro/pull/2455/commits/57ef368881d053254364742e56e270d4dad67fc6

asadali145 reviewed (approved):
LGTM!
